shared/AppInsightsCommon/src/ThrottleMgr.ts

```diff
@@ -0,0 +1,251 @@
+            _logger = safeGetLogger(core);
+            _isTriggered = false;
+            _namePrefix = isNotNullOrUndefined(namePrefix)? namePrefix : "";
+            _queue = [];
+            let configMgr = throttleMgr;
+            _config = {} as any;
+            _config.disabled = !!configMgr.disabled;
+            _config.msgKey = configMgr.msgKey;
+            // default: send data on 28th every 3 month each year
+            let interval = {
+                // dafault: sent every three months
+                monthInterval: configMgr.interval?.monthInterval || 3,
+                dayInterval : configMgr.interval?.dayInterval || 28,
+                maxTimesPerMonth: configMgr.interval?.maxTimesPerMonth || 1
+            };
+            _config.interval = interval;
+            let limit = {
+                samplingPercentage: configMgr.limit?.samplingPercentage || 100,
+                // dafault: every time sent only 1 event
+                maxSendNumber: configMgr.limit?.maxSendNumber || 1
+            };
+            _config.limit = limit;
+            _localStorageName = _getLocalStorageName(_config.msgKey, _namePrefix);
+            
+            if (_canUseLocalStorage && _localStorageName) {
+                _localStorageObj = _getLocalStorageObj(utlGetLocalStorage(_logger, _localStorageName), _logger, _localStorageName);
+            }
+            if (_localStorageObj) {
+                // will reset _isTriggered status when sendMessage is called.
+                _isTriggered = _localStorageObj.isTriggered;
+            }
+        }
+
+        function _canThrottle(config: IthrottleMgrConfig, canUseLocalStorage: boolean, localStorageObj: IthrottleLocalStorageObj) {
+            if (!config.disabled && canUseLocalStorage && isNotNullOrUndefined(localStorageObj)) {
+                let curDate = _getThrottleDate();
+                let date = localStorageObj.date;
+                let interval = config.interval;
+                let monthExpand = (curDate.getUTCFullYear() - date.getUTCFullYear()) * 12 + curDate.getUTCMonth() - date.getUTCMonth();
+                let monthCheck = _checkInterval(interval.monthInterval, 0, monthExpand);
+                let dayCheck = _checkInterval(interval.dayInterval, 0, curDate.getUTCDate()) - 1;
+                return monthCheck >= 0 && dayCheck >= 0 && dayCheck <= config.interval.maxTimesPerMonth;
+            }
+            return false;
+        }
+
+        function _getLocalStorageName(msgKey: IThrottleMsgKey, prefix?: string) {
+            let fix = isNotNullOrUndefined(prefix)? prefix : "";
+            if (msgKey) {
+                return THROTTLE_STORAGE_PREFIX + fix + "-" + msgKey;
+            }
+            return null;
+        }
+
+        // transfer local storage string value to object that identifies start date and current count
+        function _getLocalStorageObj(value: string, logger: IDiagnosticLogger, storageName: string) {
```

Comment:
I missed this previously, noticed as part of going through the tests.
lets store this as a single JSON blob, so that the "storageName" is a key into the blob. And then we can have the values in the blob as straight forward values, this then reduces the validate required a little.